File: python/src/arker/codesandbox/_sandbox_client.py
"""`SandboxClient` — returned by `Sandbox.connect()`.

Holds `.commands`, `.fs`. Mirrors `@codesandbox/sdk` SandboxClient.
Long-running primitives (`.shells`, `.tasks`, `.ports`, `.interpreters`,
`.hosts`) throw NotImplementedError — they need WS / persistent connection
plumbing we don't have.
"""

from __future__ import annotations

import logging
from typing import TYPE_CHECKING

from ._commands import Commands
from ._filesystem import Filesystem

if TYPE_CHECKING:
    from ._sandbox import Sandbox

logger = logging.getLogger("arker.codesandbox")


class _NotImplementedNamespace:
    def __init__(self, name: str, reason: str) -> None:
        self._name = name
        self._reason = reason

    def __getattr__(self, attr: str):
        raise NotImplementedError(
            f"arker.codesandbox: SandboxClient.{self._name}.{attr} is not "
            f"supported — {self._reason}"
        )


class SandboxClient:
    """Returned by `Sandbox.connect()`. Tied to a Sandbox lifetime."""

    def __init__(self, sandbox: "Sandbox") -> None:
        self._sandbox = sandbox
        self._disposed = False
        self.commands = Commands(self)
        self.fs = Filesystem(self)
        self.shells = _NotImplementedNamespace(
            "shells",
            "interactive shells need a WS client we haven't shipped",
        )
        self.tasks = _NotImplementedNamespace(
            "tasks",
            "task definitions are codesandbox-specific project metadata",
        )
        self.ports = _NotImplementedNamespace(
            "ports",
            "port previews need a persistent tunnel URL bound to the sandbox lifetime",
        )
        self.interpreters = _NotImplementedNamespace(
            "interpreters",
            "stateful interpreters need a long-running kernel",
        )
        self.hosts = _NotImplementedNamespace(
            "hosts",
            "host tokens are codesandbox auth-server primitives",
        )

    def disconnect(self) -> None:
        """Local-only cleanup. We don't hold a persistent connection."""
        self._disposed = True

File: typescript/src/codesandbox/sandbox-client.ts
import { Commands } from "./commands.js";
import { Filesystem } from "./filesystem.js";
import type { Sandbox } from "./sandbox.js";

function makeStub(name: string, reason: string): Record<string, unknown> {
  return new Proxy({}, {
    get(_target, prop) {
      throw new Error(
        `arker.codesandbox: SandboxClient.${name}.${String(prop)} is not supported — ${reason}`,
      );
    },
  });
}

/** Mirrors `@codesandbox/sdk` SandboxClient. Returned by `Sandbox.connect()`. */
export class SandboxClient {
  readonly _sandbox: Sandbox;
  readonly commands: Commands;
  readonly fs: Filesystem;
  readonly shells: Record<string, unknown>;
  readonly tasks: Record<string, unknown>;
  readonly ports: Record<string, unknown>;
  readonly interpreters: Record<string, unknown>;
  readonly hosts: Record<string, unknown>;

  constructor(sandbox: Sandbox) {
    this._sandbox = sandbox;
    this.commands = new Commands(this);
    this.fs = new Filesystem(this);
    this.shells = makeStub("shells", "interactive shells need a WS client");
    this.tasks = makeStub("tasks", "tasks are codesandbox-specific project metadata");
    this.ports = makeStub("ports", "port previews need persistent tunnels bound to sandbox lifetime");
    this.interpreters = makeStub("interpreters", "stateful interpreters need a long-running kernel");
    this.hosts = makeStub("hosts", "host tokens are codesandbox auth-server primitives");
  }

  disconnect(): void {
    /* no-op — we hold no persistent connection */
  }
}
